### PR TITLE
Add TouchBar Support

### LINKED ISSUE #7781: Support MacBook Touch Bar API
Tracking issue for adding touch bar API support to Electron on macOS 👉 ⌨️   💻  🍎 

- [Chrome issue](https://bugs.chromium.org/p/chromium/issues/detail?id=660126)
- [NSTouchBar docs](https://developer.apple.com/reference/appkit/nstouchbar?language=objc)
- [Design video](http://images.apple.com/media/us/macbook-pro/2016/b4a9efaa_6fe5_4075_a9d0_8e4592d6146c/films/design/macbook-pro-design-tft-cc-us-20161026_1536x640h.mp4)
- [NSTouchBar catalog](https://developer.apple.com/library/content/samplecode/NSTouchBarCatalog/Introduction/Intro.html)
- [Touch bar example app](https://developer.apple.com/library/content/samplecode/ToolbarSample/Introduction/Intro.html)

Please feel free to leave comments here about API ideas, usage patterns, etc.

### PR DESCRIPTION
**Pull request by MarshallOfSound**
So here goes 👍 

I've still got to add docs for all this but basically this PR adds support for

* `window.setTouchBar(bar)`
* `new TouchBar([...])`
* `new TouchBarButton(...)`
* `new TouchBarColorPicker(...)`
* `new TouchBarGroup(...)`
* `new TouchBarLabel(...)`
* `new TouchBarPopover(...)`
* `new TouchBarSpacer(...)`
* `new TouchBarSlider(...)`

Still to do

* [ ] ~Add support for the "customizable" attribute (pretty major thing missing but it requires some good docs and some magical techniques)~
* [x] Move item classes to their own files instead of being static classes on `TouchBar`
* [x] Document all the things
* [ ] ~Implement a `NSCustomTouchBarItem` with an `NSScrubber` as the view (this is quite a big task, not sure if it is even possible to declare at runtime)~
* [ ] ~Implement a ` NSCustomTouchBarItem` with an  [`NSSegmentedControl`](https://developer.apple.com/reference/appkit/nssegmentedcontrol?language=objc) as the view!~
* [x] Test? Somehow? Idk?
* [x] Ensure this doesn't break on older macOS versions. (Test on El Capitan)
* [x] API Feedback / changes
* [x] Remove modifications to default_app
* [x] Fix linting 😆 
* [x] Add methods to manually set the slider value
* [x] Add a way to create the "spacer" touchbar items

Closes #7781

### CODE REVIEW

**Comment by MarshallOfSound** [comment]
@zcbenz Given this compiles without any issue on my laptop I think the OSX build machine needs to have it's XCode updated?

**Comment by bengotow** [comment]
This looks great! Will try it today with N1. You're correct - this will need to be built with the target SDK set to MacOS 10.12.1 with the most recent version of Xcode. I wonder how often Travis updates Xcode...

**Comment by MarshallOfSound** [comment]
@bengotow It looks like travis-ci already has support for the required version of X-Code

https://docs.travis-ci.com/user/languages/objective-c/ 😄

**Comment by krisnarocks** [comment]
Would like to point this cool app,
https://red-sweater.com/touche/

It emulates TouchBar on all mac (as it claims). So, if you need to do some testing on a mac without TouchBar, you could do it now.

**Comment by MarshallOfSound** [comment]
In order to reduce API overhead (keep methods minimal) I added a method to all `TouchBarItem` instances called `updateConfig` (probably going to rename to `updateOptions`) that will update the relevant `NSTouchBarItem` instances with the new config without redrawing the entire TouchBar.  Seems to work nicely 👍

**Comment by bcoe** [comment]
@MarshallOfSound hey, I'm attempting to take this branch for a spin and am running into some issues.

* I've checked out your branch.
* built it and installed it globally (`npm i -g`).
* I've tried creating a demo app that looks like this: https://gist.github.com/bcoe/8e658eb17cc0616677a74a06e155e4d6

When I start up the project, `electron .` in the project folder; focusing on the window doesn't seem to modify the touch bar ... I'm fairly new to electron development, so please excuse my ignorance if I'm missing something simple.

**Comment by zeke** [comment]
Hi @bcoe!

**Comment by MarshallOfSound** [comment]
@bcoe This part is probably where you went wrong

> built it and installed it globally (npm i -g).

There isn't currently an easy way (that I know of) to link the version of Electron you built with an the `electron` command in another module.  Whenever I test I just copy the output of the electron build to `node_modules/electron/dist` in whatever project I want to use it.

If you launch the sample application (that is generated during the build) do you get the touch bar buttons and such?

**Comment by ktiedt** [comment]
This is amazing progress :D 

Any chance the popover can allow for options (if it doesnt already - I didn't see) to mimic the volume and or brightness popovers? Or the animated popover for the default function keys?

**Comment by MarshallOfSound** [comment]
@ktiedt I briefly looked into mirroring that kind of functionality and the TLDR is.

* I couldn't find a way to do the animation (maybe private API?, or maybe I'm blind and missed the docs?) 😆 
* You can replicate the brightness and volume sliders with some craftily placed spacers in the popover touchbar

**Comment by tvanier** [comment]
Thanks @MarshallOfSound for the awesome TouchBar support! Was wondering if someone tried a button with an image? I rebuilt the default app with the button below (in `default_app.js`), but it won't show up in the touch bar. Maybe some special image format/size is required? I could just find "2x with a maximum height of 30 points (corresponding to 60 pixels)" [in these docs](https://developer.apple.com/reference/appkit/nstouchbar#2647104).

```
const imageButton = new (TouchBar.Button)({
    image: path.join(__dirname, 'myimage@2x.png'),
    click: () => {
        console.log('my image clicked');
    }
});
```

**Comment by MarshallOfSound** [comment]
@tvanier I did while I was building out the API.  Try removing the "@2x" from the file name

**Comment by bcoe** [comment]
@MarshallOfSound thanks for your help, you're right I was not using the correct version of electron to test. My testing went pretty well; I got things running with hyper-term for a proof of concept:

https://twitter.com/BenjaminCoe/status/830607339577225218

some feedback for you:

* I found I was able to update the label on existing buttons (which was how I implemented the feature shown above).
* I was having trouble updating a label element; updating the label did not cause a redraw -- and you'd end up with draw artifacts.
* in general I'm curious if it's possible to expose more control over text layout, my goal is to:
  * show git and build information in the touchbar -- so fairly complex text output (this would rock!).

amazing work! can't wait for this to eventually land.

**Comment by MarshallOfSound** [comment]
> I was having trouble updating a label element; updating the label did not cause a redraw -- and you'd end up with draw artifacts.

Will experiment with that, I have an example locally which implements a clock in a label element (one of the local testing things I did) and it appears to update nicely.

> in general I'm curious if it's possible to expose more control over text layout, my goal is to:

The best I can do really is expose the core API's of the touch bar, anything super advanced requires custom views in native code and can't really be generalized well enough.

> show git and build information in the touchbar -- so fairly complex text output (this would rock!).

First off, that would be aswesome, just having like `Webpack Status: Building` or something in the touchbar would save a few alt tabs to watch it rebuild.

Secondly, what exactly do you need in order to achieve what you want to do?  Can't this kind of text output be done with labels?

**Comment by matthewdias** [comment]
According to the spec, the touchbar isn't supposed to be used for displaying information. It's strictly for controls.

**Comment by NTag** [comment]
@matthewdias However, when using QuickTime for a screencast, the touchbar displays the length and the size of the recording, which is pure information.

capture d ecran touch bar 2017-02-12 a 10 32 23

**Comment by jhen0409** [comment]
Just found a way to add button to Control Strip, but it's __private API__: https://github.com/a2/touch-baer

Probably only these applications are using it: Xcode, [TouchSwitcher](https://hazeover.com/touchswitcher.html), [Rocket](http://julianthayn.com/rocket)

Currently it's I feeling the best experience for use TouchBar, but I think it will break the experience of `MPRemoteCommandCenter`, this may be the reason Apple didn't public it? 🤔

**Comment by Cryrivers** [comment]
Any updates for this PR?

**Comment by PeterTeng** [comment]
The last update was 2 months ago, any one approach to this feature too?

**Comment by MarshallOfSound** [comment]
@Cryrivers @PeterTeng This PR is currently blocked by the inability to build Chromium with the required target SDK.

See https://github.com/electron/electron/pull/8280#issuecomment-273334678

**Comment by kevinsawicki** [comment]
@MarshallOfSound perhaps this can be unblocked using this technique, https://cs.chromium.org/chromium/src/ui/base/cocoa/touch_bar_forward_declarations.h

**Comment by MarshallOfSound** [comment]
@kevinsawicki Will look into next week some time (unless someone plays with it first).  Will have to downgrade my SDK and see if it still compiles 😆 What exact combinations of SDK / xCode do the build machines use again?

**Comment by kevinsawicki** [comment]
> What exact combinations of SDK / xCode do the build machines use again?

Internal CI has XCode 8.2.1 and 10.10 SDK

**Comment by galli-leo** [comment]
Would it be possible to include TouchID functionality as well? I really like https://hyper.is as terminal client and it would be very cool to implement TouchID for sudo.

**Comment by kevinsawicki** [comment]
> Would it be possible to include TouchID functionality as well?

That is outside the scope of this pull request, please open a separate issue for it, thanks.

File: lib/browser/api/touch-bar.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+class TouchBarItem extends EventEmitter {
+  constructor (config) {
+    super()
+    this.id = `${nextItemID++}`
+  }
+
+  _addLiveProperty (name, initialValue) {
+    const privateName = `_${name}`
+    this[privateName] = initialValue
```
Comment (MarshallOfSound) [inline lib/browser/api/touch-bar.js]: Can't this just be

```js
let liveValue = initialValue
Object.defineProperty(this, name, {
  get: function () {
    return liveValue
  },
  set: function (value) {
    liveValue = value
    this.emit('change')
  },
})
```

Then we aren't exposing the private variable on the `this` scope

Comment (kevinsawicki) [inline lib/browser/api/touch-bar.js]: The underscore version is used other places to set the property directly in response to an interaction event without emitting a change event so the UI isn't updated in response to an interaction.

Comment (MarshallOfSound) [inline lib/browser/api/touch-bar.js]: Ah, nice 😄

File: docs/api/touch-bar.md [hunk 2]
```diff
@@ -6,39 +6,114 @@ Process: [Main](../tutorial/quick-start.md#main-process)
...
+}
+
+const updateReels = () => {
+  reel1.label = getRandomValue()
+  reel2.label = getRandomValue()
+  reel3.label = getRandomValue()
+}
+
+const finishSpin = () => {
+  const uniqueValues = new Set([reel1.label, reel2.label, reel3.label]).size
```
Comment (zeke) [inline docs/api/touch-bar.md]: I didn't know about `Set`. Pretty cool.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

**Comment by kevinsawicki** [comment]
I've tested this on El Capitan to verify it works on older OS versions and a also machine without a touch bar.

Thanks @MarshallOfSound for your awesome work on this 👍 🎉 🥇 

It will be included in the next Electron beta release, 1.6.3.

slot-machine

**Comment by tvanier** [comment]
Nice job @MarshallOfSound and @kevinsawicki ! I will retry a button with an image file, this did not work for me a few weeks back, likely because of my png. Maybe add an example with a working image?

```
const imageButton = new (TouchBar.Button)({
    image: path.join(__dirname, 'my-icon.png'),
    click: () => {
        console.log('my image clicked');
    }
});
```

**Comment by kevinsawicki** [comment]
>  image: path.join(__dirname, 'my-icon.png'),

The property is `icon` now to match the `MenuItem` API.

**Comment by tvanier** [comment]
Ah ok, although `NSButton` uses `image` I think. I guess these icon specs should work?

> If you use an image in a button or other control in the Touch Bar, take care to employ a template image. Template images in the Touch Bar respond automatically to system white-point changes, and automatically react to user interactions.
Use Retina-resolution image assets, designated as `@2x` in your asset catalog and with a maximum height of 30 points (corresponding to 60 pixels).

https://developer.apple.com/reference/appkit/nstouchbar#2647104

**Comment by kevinsawicki** [comment]
>  I guess these icon specs should work?

Yup, they should, I've tried with this icon locally called `electron@2x.png` and it worked for me:
electron 2x

**Comment by PeterTeng** [comment]
Nice job @MarshallOfSound and @kevinsawicki ! !!!

**Comment by rhysd** [comment]
I tried master branch and this really works nice! Thank you.

**Pull request merged**